File: jamcoin/jamcoin.js
process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();    
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

function isJamcoinPrime(num){
    var max = 100;
    if (num < max) {
        max = Math.sqrt(num);
    }
    for (var divisor = 2; divisor < max; divisor++) {
        if (num % divisor == 0) {
            return divisor;
        }
    }
    return true
}

function main() {
    var tests = readLine();
    var inputs = readLine();
    inputs = inputs.split(" ");
    var jamcoinLength = parseInt(inputs[0]), jamcoinsRequired = parseInt(inputs[1]);
    
    var foundJamcoins = 0;
    var jamcoins = [], divisors = [];
    
    var currentCoin = [];
    for (var count = 0; count < jamcoinLength; count++) {
        if (count == 0 || count == jamcoinLength - 1) {
            currentCoin.push("1");
        } else {
            currentCoin.push("0");
        }
    }
    
    // we now have the string for the initial jam coin we will need the jamcoin
    // in base 10 to ensure we can add 2 to it each loop. This causes the jamcoin
    // to retain the "1" in the final position
    var jamcoinBase10 = parseInt(currentCoin.join(''), 2)
    // console.log(jamcoinBase10);
    var primeFlag = false
    
    while (foundJamcoins < jamcoinsRequired) {
        // to find the jam coin we will need to test each coins value under each
        // base binary, ternary, quaternary, quinary, senary, septenary, octonary
        // nonary and decimal to ensure the value is not a prime number
        // ensure the current divisors array contains an empty arry
        divisors[foundJamcoins] = [];
        jamcoins[foundJamcoins] = currentCoin.join('')
        for (var base = 2; base <= 10 && !primeFlag; base++) {
            var coinValue = parseInt(currentCoin.join(''), base);
            var result = isJamcoinPrime(coinValue);
            // console.log("coinValue: " + coinValue + " | result: " + result);
            if (result === true) {
                // this coin value is prime. stop checking this coin.
                primeFlag = true;
                jamcoins[foundJamcoins] = "";
                divisors[foundJamcoins] = [];
            } else {
                // coinValue not prime. jamcoin still viable add the divisor returned by
                // isJamcoinPrime to the divisors array
                divisors[foundJamcoins][base - 2] = result;
            }
        }
        // all bases checked verify the prime flag
        if (!primeFlag) {
            foundJamcoins++;
        }
        // increment the jamcoin string by 2
        jamcoinBase10 += 2;
        // console.log(jamcoinBase10)
        currentCoin = jamcoinBase10.toString(2).split('');
        // console.log(currentCoin.join(''))
        primeFlag = false;
    }
    
    console.log("Case #1:");
    for (var coin = 0; coin < jamcoinsRequired; coin++) {
        var output = "";
        output += jamcoins[coin];
        for (var base = 2; base <= 10; base++) {
            output += " " + divisors[coin][base - 2]
        }
        console.log(output);
    }
}

File: jamcoin/jamcoin.rb
#first need a way to check for prime number
def is_prime?(num)
  return false if num <= 1
  2.upto(1000).each {|i| return i if num % i == 0}
  true
end

input = gets.chomp.to_i

input = gets.chomp.split(" ")

jam_length = input[0].to_i
jam_count = input[1].to_i
found_jam_coins = 0
jamcoins = []
divisors = [] 

current_coin = []
[*0...jam_length].each do |count|
    if count == 0 || count == jam_length - 1
        current_coin.push("1")
    else
        current_coin.push("0")
    end
end
#we now have the base 10 version of the coin. this will help us to get
#the next version of the coin each time by adding 2 to the number
jamcoinbase10 = current_coin.join('').to_i(2)
prime_flag = false

while found_jam_coins < jam_count do
    divisors[found_jam_coins] = []
    jamcoins[found_jam_coins] = current_coin.join('')
    # current_coin.split('')
    [*2..10].each do |base| 
        if !prime_flag
            coinvalue = current_coin.join('').to_i(base)
            result = is_prime?(coinvalue)
            if result === true
                #this value is prime and as such not a jamcoin set 
                prime_flag = true
                jamcoins[found_jam_coins] = ""
                divisors[found_jam_coins] = []
            else
                # value is not prime and still a possible jamcoin
                divisors[found_jam_coins][base - 2] = result
            end
            #loop around will chech the next base
        end
    end
    # all bases checked. look at the prime flag
    if !prime_flag
        #this is a valid coin since no primes were found increase the found jam_coins
        found_jam_coins += 1
    end
    #move onto the next possible coin. do this by adding 2 to the jamcoinbase10
    jamcoinbase10 += 2
    #convert this into a binary string
    current_coin = jamcoinbase10.to_s(2).split('')
    #reset prime flag
    prime_flag = false
end

#should have found all jam coins lets print

puts "Case #1:"
[*0...jam_count].each do |index|
    output = ""
    output = "#{jamcoins[index]}"
    divisors[index].each do |divisor|
        output += " #{divisor}"
    end
    puts output
end